toonz/sources/toonz/sceneviewer.cpp

```diff
@@ -1715,12 +1719,9 @@ void SceneViewer::GLInvalidateAll() {
 //-----------------------------------------------------------------------------
 
 void SceneViewer::GLInvalidateRect(const TRectD &rect) {
-  // there is a case that this function is called more than once before
-  // paintGL() is called
-  if (!m_clipRect.isEmpty())
-    m_clipRect += rect;
-  else
-    m_clipRect = rect;
+  if (m_dirty) m_clipRect += rect;
+          else m_clipRect  = rect;
+  m_dirty = true;
```

Comment:
> When we called SceneViewer::GLInvalidateAll() and after that SceneViewer::GLInvalidateRect(), then invalidation rectangle will reduced. This PR fixes it.

I understand that such scenario will define the `m_clipRect` and will fail to update whole viewer. Good catch.

However I wonder whether this fix will resolve it.
In the `GLInvalidateRect()`, regardless of `m_dirty` value `m_clipRect` seems to be set to some value , which will activate clipping anyway.

Instead, I think you need to set `m_dirty` to true ONLY when `GLInvalidateAll()` is called, and if `m_dirty` is true `m_clipRect` should be kept empty in `GLInvalidateRect()` in order to update whole viewer.

Reply:
Calling GLInvalidateRect with empty rectangle should be equal to invalidate all. This is behavior of previous versions of GLInvalidateRect function.

I think the valid solution is:

```
void SceneViewer::GLInvalidateRect(const TRectD &rect) {
  if (!m_dirty || rect.isEmpty())
    m_clipRect = rect;
  else if (!m_clipRect.isEmpty())
    m_clipRect += rect;
  m_dirty = true;
  update();
  if (m_vRuler) m_vRuler->update();
  if (m_hRuler) m_hRuler->update();
}
```

Or should we change behavior GLInvalidateRect to consider empty rectangle as empty instead of infinite?